### PR TITLE
Individual support for parsing of VFPPROD tables

### PR DESCRIPTION
**Pull request by babrodtk**
This runs on my setup, and places the data in the correct variables inthe VFPProdTable class. It should do nothing if there are no VFPPROD keywords in a deck, and only adds functionality.

What is missing here, is the support for units, which will come in a subsequent PR.
Edited: Units should now be supported

### CODE REVIEW

File: opm/parser/eclipse/EclipseState/EclipseState.hpp [hunk 1]
```diff
@@ -66,6 +66,7 @@
 #include <set>
 #include <memory>
 #include <iostream>
+#include "Tables/VFPProdTable.hpp"
```
Comment (atgeirr) [inline opm/parser/eclipse/EclipseState/EclipseState.hpp]: The include should be of the same form as the table includes above (and with them).

File: opm/parser/eclipse/EclipseState/EclipseState.hpp [hunk 2]
```diff
@@ -220,6 +222,9 @@ namespace Opm {
         void initPlyshlogTables(DeckConstPtr deck,
                                               const std::string& keywordName,
                                               std::vector<PlyshlogTable>& tableVector);
+        void initVFPProdTables(DeckConstPtr deck,
+                                              const std::string& keywordName,
```
Comment (atgeirr) [inline opm/parser/eclipse/EclipseState/EclipseState.hpp]: Strange indentation here, I was about to ask that you check that you are using spaces (and not tabs), but then I saw that the indentation of the method above is also a little off...

File: opm/parser/eclipse/EclipseState/Tables/VFPProdTable.hpp [hunk 3]
```diff
@@ -0,0 +1,536 @@
...
+    void init(DeckKeywordConstPtr table) {
+        auto iter = table->begin();
+
+        auto header = (*iter++);
+
+        assert(itemValid(header, "TABLE"));
+        m_table_num = header->getItem("TABLE")->getInt(0);
+
+        assert(itemValid(header, "DATUM_DEPTH"));
+        m_datum_depth = header->getItem("DATUM_DEPTH")->getRawDouble(0);
```
Comment (atgeirr) [inline opm/parser/eclipse/EclipseState/Tables/VFPProdTable.hpp]: Should probably be using getSIDouble()?

**Comment by joakim-hove** [comment]
Thank you - I appreciate this code targetting opm-parser. Unfortunately I don't have possibility to review it properly today.

File: opm/parser/eclipse/EclipseState/Tables/VFPProdTable.hpp (hunk 3 above)
Comment (babrodtk) [inline opm/parser/eclipse/EclipseState/Tables/VFPProdTable.hpp]: Yes, the whole units-thing is sort of missing from this PR.

File: opm/parser/eclipse/EclipseState/EclipseState.hpp (hunk 1 above)
Comment (babrodtk) [inline opm/parser/eclipse/EclipseState/EclipseState.hpp]: Thanks for spotting this: Fixing

File: opm/parser/eclipse/EclipseState/EclipseState.cpp [hunk 4]
```diff
@@ -652,6 +658,21 @@ namespace Opm {
...
     }
 
+    void EclipseState::initVFPProdTables(DeckConstPtr deck,
+                                          const std::string& keywordName,
+                                          std::vector<VFPProdTable>& tableVector) {
+        if (!deck->hasKeyword(keywordName)) {
+            return;
+        }
+
+        int num_tables = deck->numKeywords("VFPPROD");
```
Comment (babrodtk) [inline opm/parser/eclipse/EclipseState/EclipseState.cpp]: Note to self:
deck->numKeywords(ParserKeywords::VFPPROD::KeywordName);

Comment (babrodtk) [inline opm/parser/eclipse/EclipseState/EclipseState.cpp]:
```C++
auto vfptables = deck->getKeywordList<ParserKeywords::VFPPROD>();
for (auto table : vfptables) {
}
```

File: opm/parser/eclipse/EclipseState/Tables/VFPProdTable.hpp [hunk 5]
```diff
@@ -0,0 +1,655 @@
...
+    void init(DeckKeywordConstPtr table, std::shared_ptr<Opm::UnitSystem> deck_unit_system) {
+        auto iter = table->begin();
+
+        auto header = (*iter++);
+
+        assert(itemValid(header, "TABLE"));
+        m_table_num = header->getItem("TABLE")->getInt(0);
+
+        assert(itemValid(header, "DATUM_DEPTH"));
+        m_datum_depth = header->getItem("DATUM_DEPTH")->getSIDouble(0);
```
Comment (babrodtk) [inline opm/parser/eclipse/EclipseState/Tables/VFPProdTable.hpp]:
```
header->getItem<ParserKeywords::VFPPROD::DATUM_DEPTH>()->getSIDouble(0)
```

**Comment by joakim-hove** [comment]
Apart from the minor comment about the unnecessary `getNonEmptyItem()` method I am ready to merge this.

**Comment by babrodtk** [comment]
Removed the getNonEmptyItem() function as requested. I agree that it appears superfluous.

**Comment by joakim-hove** [comment]
Good work - merging.

File: opm/parser/eclipse/EclipseState/Tables/tests/TableTests.cpp [hunk 6]
```diff
@@ -355,3 +357,686 @@ BOOST_AUTO_TEST_CASE(PvtoTable_Tests) {
...
+        Opm::ParserPtr parser(new Opm::Parser);
+        Opm::DeckConstPtr deck(parser->parseString(missing_metadata));
+        Opm::DeckKeywordConstPtr vfpprodKeyword = deck->getKeyword("VFPPROD");
+        std::shared_ptr<Opm::UnitSystem> units(Opm::UnitSystem::newMETRIC());
+        BOOST_CHECK_EQUAL(deck->numKeywords("VFPPROD"), 1);
+
+        Opm::VFPProdTable vfpprodTable;
+
+
+        BOOST_CHECK_THROW(vfpprodTable.init(vfpprodKeyword, units), std::invalid_argument);
```
Comment (atgeirr) [inline opm/parser/eclipse/EclipseState/Tables/tests/TableTests.cpp]: This check fails for me: it throws
```
std::out_of_range: Index must be smaller than 0 but is 0
```
from something called from line 87 of VFPProdTable.cpp:
```
m_flo_type = getFloType(header->getItem<VFPPROD::RATE_TYPE>()->getString(0));
```
whereas the check specifies an std::invalid_argument.

Comment (babrodtk) [inline opm/parser/eclipse/EclipseState/Tables/tests/TableTests.cpp]: You're absolutely right. The cause is the removal of the getNonEmptyItem() function which instead threw the std::invalid_argument

**Pull request merged**